src/obfuscator.ts

```diff
@@ -0,0 +1,267 @@
+/// <reference path="./typings/tyranid-extensions.d.ts" />
+
+import { Tyr } from "tyranid";
+import { Cursor, Collection, FilterQuery, BulkWriteResult, UpdateWriteOpResult, UnorderedBulkOperation } from "mongodb";
+import { Timestamp, ObjectId, ObjectID } from 'bson';
+import { AesUtil } from './encryptor';
+import { emit } from "cluster";
+
+const defaultBatchSize = 500;
+const idOnlyProjection = { _id: 1 };
+interface IdRecord {
+  _id: ObjectID
+}
+interface FieldObject {
+  [name: string]: string;
+}
+
+/**
+ * TODO: Really large Document support with GridFS API
+ */
+Tyr.obfuscate = async (opts: Tyr.ObfuscateBatchOpts): Promise<Tyr.ObfuscateBatchResult> => {
+  const { collection, query, replacementValues, replacementValCollection, metadataSuffix } = opts;
+
+  const sourceCollectionName: string = collection.def.dbName;
+  const mongoSrcCollection: Collection = Tyr.db.collection(sourceCollectionName);
+
+  // Complicated extending Tyr.Collection to have configurable name
+  // Going with convention for now
+  const metaCollecName = `${sourceCollectionName}${metadataSuffix}`;
+  const metaDataCollection: Collection = await getMetaDataCollection(metaCollecName);
+  
+  const obfsctFields = obfuscateableFieldsFromCollection(collection);
+
+  return await createBatchMetaData(obfsctFields, mongoSrcCollection, metaDataCollection, query, replacementValues, replacementValCollection);
```

Comment:
Nit: Don't think you need the `await` since you are already returning the Promise from `createBatchMetaData`. Might save some generated code/perf for platforms without native `await` yet.